rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport.h

```diff
@@ -35,25 +252,555 @@ namespace rmw_fastrtps_cpp
+                            throw std::runtime_error("string overcomes the maximum length");
+                        }
+                        ser << str;
+                    }
+                    break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE:
+                    {
+                        const MembersType *sub_members = (const MembersType*)member->members_->data;
+                        serializeROSmessage(ser, sub_members, field);
+                    }
+                    break;
+                default:
+                    printf("unknown type id %u\n", member->type_id_);
+                    throw std::runtime_error("unknown type");
+            }
+        }
+        else
+        {
+            switch(member->type_id_)
+            {
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BOOL:
+                    SER_ARRAY_SIZE_AND_VALUES(bool)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_BYTE:
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT8:
+                    SER_ARRAY_SIZE_AND_VALUES(uint8_t)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_CHAR:
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT8:
+                    SER_ARRAY_SIZE_AND_VALUES(char)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_FLOAT32:
+                    SER_ARRAY_SIZE_AND_VALUES(float)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_FLOAT64:
+                    SER_ARRAY_SIZE_AND_VALUES(double)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT16:
+                    SER_ARRAY_SIZE_AND_VALUES(int16_t)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT16:
+                    SER_ARRAY_SIZE_AND_VALUES(uint16_t)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT32:
+                    SER_ARRAY_SIZE_AND_VALUES(int32_t)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT32:
+                    SER_ARRAY_SIZE_AND_VALUES(uint32_t)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_INT64:
+                    SER_ARRAY_SIZE_AND_VALUES(int64_t)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_UINT64:
+                    SER_ARRAY_SIZE_AND_VALUES(uint64_t)
+                        break;
+                case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING:
+                    SER_ARRAY_SIZE_AND_VALUES(std::string)
```

Comment:
You probably need logic here to select between `std::string` and `rosidl_generator_c__String`.